Spatial_searching/include/CGAL/Splitters.h

```diff
@@ -303,5 +303,179 @@ namespace CGAL {
+
+    Balanced_splitter() : Base() { }
+    Balanced_splitter(const unsigned int bucket_size) : Base(bucket_size) { }
+
+  private:
+    void print_reference(
+      const std::size_t dim,
+      const std::vector<FTP>& reference) const {
+
+      std::cout << std::endl;
+      for (std::size_t i = 0; i < reference.size(); ++i) {
+        std::cout << *(reference[i]) << std::endl;
+        for (std::size_t k = 0; k < dim; ++k) {
+          std::cout << reference[i][k] << " ";
+        }
+        std::cout << std::endl;
+      }
+    }
+
+    void print_references(
+      const std::size_t dim,
+      const std::vector< std::vector<FTP> >& references,
+      const std::string name) const {
+
+      std::cout << std::endl << "--- " << name << " : " << std::endl;
+      for (const auto& reference : references) {
+        print_reference(dim, reference);
+      }
+    }
+
+  public:
+    void operator()(
+      Separator& sep, Container& c0, Container& c1) const {
+
+      CGAL_assertion(c0.is_valid());
+      CGAL_assertion(c1.is_valid());
+
+      CGAL_assertion(!c0.is_last_call());
+      CGAL_assertion(!c1.is_last_call());
+
+      CGAL_assertion(c0.dimension() == c1.dimension());
+
+      const std::size_t dim = c0.get_dim();
+      const std::size_t depth = c0.get_depth();
+
+      const long start = c0.get_start();
+      const long end = c0.get_end();
+
+      const auto& ref_ptr = c0.get_ref_ptr();
+      auto& references = *(ref_ptr);
+
+      #ifdef KD_TREE_DEBUG
+      std::cout << std::endl;
+      std::cout << "- dim: "   << dim   << std::endl;
+      std::cout << "- depth: " << depth << std::endl;
+      std::cout << "- start: " << start << std::endl;
+      std::cout << "- end: "   << end   << std::endl;
+      #endif
+
+      CGAL_assertion(dim   != static_cast<std::size_t>(-1));
+      CGAL_assertion(depth != static_cast<std::size_t>(-1));
+
+      CGAL_assertion(start != -1);
+      CGAL_assertion(end   != -1);
+
+      const long median = start + ((end - start) / 2);
+      const std::size_t axis = depth % dim;
+      CGAL_assertion(median >= 0);
+
+      #ifdef KD_TREE_DEBUG
+      std::cout << "- median: " << median << std::endl;
+      #endif
+
+      const Key_compare compare_keys;
+      std::vector<FTP> tmp(references[0].size());
+      CGAL_assertion_msg(end > start + 2, "ERROR: LAST CALL MUST BE AVOIDED!");
+
+      // Copy the first ref to save it.
+      for (long i = start; i <= end; ++i) {
+        tmp[i] = references[0][i];
+      }
+
+      // Apply permutations for dim - 1 axis.
+      long lower, upper;
+      for (std::size_t k = 1; k < dim; ++k) {
+        lower = start - 1;
+        upper = median;
+
+        for (long i = start; i <= end; ++i) { // always skip median key
+          const FT compare = compare_keys(
+            references[k][i], references[0][median], axis, dim);
+          if (compare < FT(0)) {
+            references[k-1][++lower] = references[k][i];
+          } else if (compare > FT(0)) {
```

Comment:
I thought it cannot be zero.

Reply:
It can.

Reply:
And  do you use `compare` as a `FT`  other than for comparing with zero?  If not use ` Comparison_result`.  And don't perform an `operator-()`.  And in case you really have to use a `FT`, then call functions such as [` is_zero()`](https://doc.cgal.org/latest/Algebraic_foundations/group__PkgAlgebraicFoundationsRef.html#ga997aad1528e7e744e54a807caa2be56e)

Reply:
Done.